### PR TITLE
fix(app): Fix patch creation for Applications with ValuesObject fields #15126

### LINKED ISSUE #15126: Error constructing app status patch: unable to find api field in struct RawExtension for the json field \"autoscaling\"
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [ ] I've pasted the output of `argocd version`.

**Describe the bug**

We've switched some applications to the new `valuesObject` syntax, and those apps have gotten "stuck". They go into a refreshing status and never finish

**To Reproduce**

Not clear to us how to reproduce. We deleted the entire `.status` of an affected application and it seemed to start working again. Not left it long enough yet to determine if it'll break again

**Expected behavior**

**Screenshots**

**Version**

```shell
argocd-server: v2.8.0+804d4b8
  BuildDate: 2023-08-07T14:25:33Z
  GitCommit: 804d4b8ca6bc4c2cf02c5c971aa923ec5b8623f0
  GitTreeState: clean
  GoVersion: go1.20.6
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v5.1.0 2023-06-19T16:58:18Z
  Helm Version: v3.12.1+gf32a527
  Kubectl Version: v0.24.2
  Jsonnet Version: v0.20.0
```

**Logs**

```
{"application":"argocd/istio-gateway-iap","level":"error","msg":"Error constructing app status patch: unable to find api field in struct RawExtension for the json field \"service\"","time":"2023-08-21T09:01:10Z"}
{"application":"argocd/istio-gateway-public","level":"error","msg":"Error constructing app status patch: unable to find api field in struct RawExtension for the json field \"service\"","time":"2023-08-21T09:01:10Z"}
{"application":"argocd/istio-gateway-vpn","level":"error","msg":"Error constructing app status patch: unable to find api field in struct RawExtension for the json field \"service\"","time":"2023-08-21T09:01:25Z"}
{"application":"argocd/istio-gateway-public","level":"error","msg":"Error constructing app status patch: unable to find api field in struct RawExtension for the json field \"service\"","time":"2023-08-21T09:01:25Z"}
{"application":"argocd/istio-gateway-vpn","level":"error","msg":"Error constructing app status patch: unable to find api field in struct RawExtension for the json field \"service\"","time":"2023-08-21T09:01:39Z"}
{"application":"argocd/istio-gateway-public","level":"error","msg":"Error constructing app status patch: unable to find api field in struct RawExtension for the json field \"service\"","time":"2023-08-21T09:01:40Z"}
```

### PR DESCRIPTION
**Pull request by PaulSonOfLars**
Fix patch creation for Applications using ValuesObject fields, and add multiple tests to validate that the fix works as expected.

Tried to cherry pick and retain commit attribution @vladfr, but that might not have worked out well. Feel free happy to rebase onto a cleaner git history if you prefer, my intention wasnt to take ownership - credits go to your fix.

Extends and tests #15227, Fixes #15126

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [X] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [X] Does this PR require documentation updates?
* [X] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [X] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [X] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by blakepettersson** [comment]
@PaulSonOfLars I don't have any specific comments about the code itself, but could you add a few e2e tests to verify that the patching of `valuesObject` works as intended? Specifically the parts in regards to the infinite refresh.

**Comment by PaulSonOfLars** [comment]
Hi @blakepettersson, thanks for your comment! I've gone ahead and added an E2E test as requested.

I chose to add the test on application sets, since they deal with apps, as well as updating apps - happy to add a more specific application-only one if you'd like (or if theres anything else?). Took me a while to get familiar with the testsuite but I got there in the end!

Also found a cleaner refactor for the merge patch along the way which covers more cases and is easier to extend, so worked out well.

**Comment by mgledi** [comment]
@blakepettersson This problem is becoming more and more severe in our production system. It would be awesome if it is merged and released soon.

**Comment by eagleusb** [comment]
> @blakepettersson This problem is becoming more and more severe in our production system. It would be awesome if it is merged and released soon.

It's the same for us, I was thinking it would be merged earlier (i.e. not forking the project) as it breaks probably a lot of applications.

**Comment by PaulSonOfLars** [comment]
I'm looking to add one more test (hopefully next week) as a sanity check, and then this should be ready to merge from my perspective. 

If I could get a review before then, I'll make sure to address any comments at the same time as the test - @blakepettersson, @crenshaw-dev?
Also, if there's anything I should focus on, or anything I can do to help speed up reviews, just let me know!

**Comment by blakepettersson** [comment]
@PaulSonOfLars LGTM IMO, pending any further tests you need to add on your end.

**Comment by PaulSonOfLars** [comment]
Perfect @blakepettersson - test has been added. All good to merge on my end.

Was bothering me that I was e2e testing appsets, but not testing apps directly; that itch has now been handled!

**Comment by vladfr** [comment]
@PaulSonOfLars really nice job on this!

File: test/e2e/app_management_test.go
Comment (blakepettersson) [inline test/e2e/app_management_test.go:527]: Could you add assertions for verifying that the app has properly synced? Something like (haven't tested, but think this should work) 
```suggestion
		Then().
		Expect(Success("")).
		Expect(SyncStatusIs(SyncStatusCodeSynced)).
		Expect(NoConditions()).
```

File: test/e2e/applicationset_test.go
Comment (blakepettersson) [inline test/e2e/applicationset_test.go:679]: Could you add assertions for verifying that the app has properly synced? Something like (haven't tested, but think this should work) 
```suggestion
		Then().
		Expect(Success("")).
		Expect(SyncStatusIs(SyncStatusCodeSynced)).
		Expect(NoConditions()).
```

Comment (PaulSonOfLars) [reply test/e2e/applicationset_test.go:679]: Application sets don't have these expectations; those are all app-only. The existing expectation of "app exists" should be enough though.

Comment (mgledi) [reply test/e2e/applicationset_test.go:679]: @blakepettersson Can you merge it? So that it is in the next release?

Comment (blakepettersson) [reply test/e2e/applicationset_test.go:679]: @mgledi I cannot - I do not have merge rights on that part of the repository

Comment (mgledi) [reply test/e2e/applicationset_test.go:679]: @crenshaw-dev What about you?

**Comment by KAllan357** [comment]
Any updates on this getting merged?

**Comment by 7onn** [comment]
Shaun The Sheep

**Comment by rsnyman** [comment]
I just ran into this issue today, and it's blocking a deployment. I would love to see this merged.

**Comment by hshepherd** [comment]
Thank you for the fix! I am seeing this issue as well. Looking forward this being released :)

**Review by tamsky — APPROVED** [review]
LGTM.

@argoproj/argocd-approvers ?

**Comment by crenshaw-dev** [comment]
I'm not fundamentally opposed to the hackaround, but since we apparently made a mistake by using the RawExtension type in the first place I want to give @PaulSonOfLars time to investigate alternatives before we merge a bunch of hack code. Paul, what are your thoughts? Merge this and fix the underlying issue later, or wait a bit to investigate some more?

**Comment by eyearian** [comment]
Can we get this merged?

**Comment by lexfrei** [comment]
@crenshaw-dev @PaulSonOfLars any news/eta on this?

**Comment by PaulSonOfLars** [comment]
@crenshaw-dev Alright, after various attempts to move away from the `runtime.RawExtension` type, I'm going to have to admit defeat here. As it stands, `strategicpatch` simply does not support the unstructured data that valuesObject requires. 

I'd recommend merging this PR for now, and we can do the move in the future if someone figure it out. 
I'll also add that if we chose to add other unstructured fields in the future (eg, kustomize patches?), the methods changed in this PR will need updating to include those too. I'll leave my other branches available (can see some attempts in #17725) in case someone tries to pick this up again in the future.

For the curious: It comes down to the fact that all strategicpatch merges expect to be done against a typed struct, and a freeform YAML/JSON object can't have that (by definition!). If the upstream k8s libraries exposed an option to fallback to standard json merge patches when no struct was found (or we hit a json.RawMessage type?), we might be able to use that; that would likely be the best path forward.

**Comment by crenshaw-dev** [comment]
Cool, let's ship it. I'm super hesitant to ship this with 2.11 since it's a change in such a core piece of code. I'm going to hold it for 2.12.

**Review by crenshaw-dev — APPROVED** [review]

**Comment by alexmt** [comment]
I want to propose another way to fix this: https://github.com/argoproj/argo-cd/pull/18061

The main reason IMHO it is safer. Less code changes and applicable only to `valuesObject`:

image


Another reason is I think the changes in `valuesObject` should be handled the same as changes in `values` which is replaced. The `valuesObject` was introduced as a convenience for `values` to avoid nested YAML/JSON. It would be strange to use merge for one and replace for another.

**Review by jessesuen — CHANGES_REQUESTED** [review]
Adding my rejection here since I see that auto-merge is already enabled for this PR, but the current thinking is to do this alternatively with: https://github.com/argoproj/argo-cd/pull/18061

**Comment by phyzical** [comment]
just thought id add, i just started running into this. was booleans as investigated. and the fix referenced worked for me a soon as i updated
👍

**Comment by blakepettersson** [comment]
Closing since this has been addressed with #18061 and #18515. Thanks for the work put in!

**Pull request closed without merging**
